pkg/operator/ceph/cluster/mgr/dashboard.go

```diff
@@ -110,58 +105,72 @@ func (c *Cluster) configureDashboardModules() error {
+func (c *Cluster) configureDashboardModuleSettings() (bool, error) {
 
-	daemonID = fmt.Sprintf("mgr.%s", daemonID)
+	monStore := config.GetMonStore(c.context, c.clusterInfo)
 
 	// url prefix
-	hasChanged, err := monStore.SetIfChanged(daemonID, "mgr/dashboard/url_prefix", c.spec.Dashboard.URLPrefix)
+	hasChanged, err := monStore.SetIfChanged(config.MgrType, "mgr/dashboard/url_prefix", c.spec.Dashboard.URLPrefix)
 	if err != nil {
 		return false, err
 	}
 
 	// ssl support
 	ssl := strconv.FormatBool(c.spec.Dashboard.SSL)
-	changed, err := monStore.SetIfChanged(daemonID, "mgr/dashboard/ssl", ssl)
+	changed, err := monStore.SetIfChanged(config.MgrType, "mgr/dashboard/ssl", ssl)
 	if err != nil {
 		return false, err
 	}
 	hasChanged = hasChanged || changed
 
 	// Prometheus host end point
 	prometheusEndpoint := c.spec.Dashboard.PrometheusEndpoint
-	changed, err = monStore.SetIfChanged(daemonID, "mgr/dashboard/PROMETHEUS_API_HOST", prometheusEndpoint)
+	changed, err = monStore.SetIfChanged(config.MgrType, "mgr/dashboard/PROMETHEUS_API_HOST", prometheusEndpoint)
 	if err != nil {
 		return false, err
 	}
 	hasChanged = hasChanged || changed
 
 	// Prometheus host end point ssl verify
 	prometheusEndpointSSLVerify := strconv.FormatBool(c.spec.Dashboard.PrometheusEndpointSSLVerify)
-	changed, err = monStore.SetIfChanged(daemonID, "mgr/dashboard/PROMETHEUS_API_SSL_VERIFY", prometheusEndpointSSLVerify)
+	changed, err = monStore.SetIfChanged(config.MgrType, "mgr/dashboard/PROMETHEUS_API_SSL_VERIFY", prometheusEndpointSSLVerify)
 	if err != nil {
 		return false, err
 	}
 	hasChanged = hasChanged || changed
 
 	// server port
 	port := strconv.Itoa(c.dashboardInternalPort())
-	changed, err = monStore.SetIfChanged(daemonID, "mgr/dashboard/server_port", port)
+	changed, err = monStore.SetIfChanged(config.MgrType, "mgr/dashboard/server_port", port)
 	if err != nil {
 		return false, err
 	}
 	hasChanged = hasChanged || changed
 
 	// SSL enabled. Needed to set specifically the ssl port setting
 	if c.spec.Dashboard.SSL {
-		changed, err = monStore.SetIfChanged(daemonID, "mgr/dashboard/ssl_server_port", port)
+		changed, err = monStore.SetIfChanged(config.MgrType, "mgr/dashboard/ssl_server_port", port)
 		if err != nil {
 			return false, err
 		}
 		hasChanged = hasChanged || changed
 	}
 
+	// Remove any existing per mgr-daemon configuration
+	if removeMgrDaemonConfiguration {
+		logger.Info("Removing any previous per mgr-daemon configuration")
+		for _, daemonID := range c.getDaemonIDs() {
+			mgrDaemonID := fmt.Sprintf("%s.%s", config.MgrType, daemonID)
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/url_prefix")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/ssl")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/PROMETHEUS_API_HOST")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/PROMETHEUS_API_SSL_VERIFY")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/server_port")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/ssl_server_port")
```

Comment:
@BlaineEXE good point, in the first version of the code I was calling this always but then I changed that because of performance concerns. As of the current code, in case we fail to delete some parameter it will be still in use as it has more precedence than the global mgr section, in the worst case this could lead to a mismatch in the configuration. Maybe adding an error log at least could be helpful to detect potential issues.

Reply:
> it has more precedence than the global mgr section, in the worst case this could lead to a mismatch in the configuration.

I think it may be good to re-reconcile if there are errors given this. Or at minimum, we should not set the "I have already completed this task" boolean that Travis mentioned if there are errors. It may be okay to let this fail with error messages, but we do want to make sure Rook is trying to get to the ideal end state as much as possible so that we don't make it an issue for users where they need to do manual steps.

Reply:
@BlaineEXE I updated the code so the new version checks whether all the delete operations were successful or not before setting the `removeMgrDaemonConfiguration` to false. Just to add that as part of the testing I observed that `ceph config rm` is really checking whether the key exists or not. It only cares about the section it self:

**success even when the key is not valid**
```
bash-4.4$ ceph config rm mgr mgr/invalid_key
bash-4.4$ echo $?
0
```
**success even when the mgr daemon doesn't exist  (mgr.X)**
```
bash-4.4$ ceph config rm mgr.X mgr/dashboard/PROMETHEUS_API_HOST
bash-4.4$ echo $?
0

```
**fails when the section is not valid**
```
bash-4.4$ ceph config rm invalid_section mgr/
Error EINVAL: unrecognized config target 'invalid_section'
```

In summary, in practice the delete will fail in case of timeout or a similar error. The ceph config is not detecting issues with invalid keys. It only detects invalid sections (not subsections such as `mgr.X`).

Reply:
This doesn't take into account what happens if Ceph is having a connection issue during the time where these commands are being sent or if there is a bug in Ceph.

Reply:
Got it! That's good investigation. It's good to make sure that the command is idempotent on Ceph's side. :)